techtonica-projects/contact-app/server/server.js

```diff
@@ -0,0 +1,126 @@
+const express = require('express');
+const cors = require('cors');
+require('dotenv').config();
+const path = require('path');
+const db = require('./db/db-connection.js');
+
+
+const app = express();
+const PORT = process.env.PORT || 8080;
+app.use(cors());
+app.use(express.json());
+
+//------------------------------------------CRUD Operations for 'contacts' & 'locations' tables----------------------------//
+
+//enppoint for full contact list
+app.get('/contacts', async (req, res) => {
+    try {
+        const { rows: contacts } = await db.query('SELECT id, name, email, phone FROM contacts');
+        res.send(contacts);
+    } catch (e) {
+        return res.status(400).json({ e });
+    }
+});
+
+//endpoint for indivual contact with details by id (join query)
+app.get('/contacts/:id', async (req, res) => {
+    try {
+        const { id } = req.params;
+        const query = `
+            SELECT contacts.id, contacts.name, contacts.email, contacts.phone, contacts.notes,
+                   locations.address, locations.city, locations.state
+            FROM contacts
+            LEFT JOIN locations ON contacts.id = locations.contact_id
+            WHERE contacts.id = $1;
+        `;
+        const result = await db.query(query, [id]);
+        if (result.rows.length === 0) {
+            return res.status(404).json({ message: 'Contact not found' });
+        }
+        res.json(result.rows[0]);
+    } catch (e) {
+        return res.status(400).json({ error: 'Something went wrong, please try again.' });
```

Comment:
We could try making the error message bit more informative by
- mentioning which of CRUD operation failed
- including one/more properties of 'e' to return more error context in json response.

Reply:
```
  return res.status(400).json({ 
            error: 'Something went wrong, please try again.',
            message: e.message,
            operation: 'PUT /contacts/:id'
        });
```

Reply:
Updated piece of code looks great, thanks for implementing. Some background for your future reference: its useful for server side error responses in full-stack application to be verbose because it helps app developer quickly figure answers to two important questions:
1. Whether issue lies on client side or server side when user errors happen.
2. Which server side API failed and what were the runtime user inputs/values that caused it. This supports developer's troubleshooting process and allows them to repro (short for reproduce) the issue by repeating user steps and intentionally giving problem values that would cause it. When developer implements a fix, the repro steps can be converted into a test case and included in the main test suite. The test case passing assures code reviewers of the change that actual bug was successfully fixed and the issue has been mitigated.